### ISSUE TITLE
MO2 Failing to Run (Flatpak)

### ISSUE THREAD

**Issue by cohesive-flight** [sonic2kk/steamtinkerlaunch#540]
Version: 10.0.20220725
OS: Pop!_OS 22.04 LTS

When attempting to run MO2 on first launch of this tool, the terminal running Flatpak Steam outputs: 

```
pid 4026 != 521, skipping destruction (fork without exec?)
/app/utils/steamtinkerlaunch/bin/steamtinkerlaunch: line 13015: /home/user/.local/share/applications/ModOrganizer-steamtinkerlaunch-dl.desktop: No such file or directory
Wed Jul  6 02:14:43 AM EDT 2022 INFO - startMO2 - Starting '/home/user/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding/MO2/ModOrganizer.exe' in standalone mode with command:
Wed Jul  6 02:14:43 AM EDT 2022 INFO - startMO2 - WINEDEBUG="-all" WINEPREFIX="/home/user/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/mo2/compatdata/pfx" "" "/home/user/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding/MO2/ModOrganizer.exe"
```

Then any subsequent launches until restarting outputs:

```
Wed Jul  6 02:16:44 AM EDT 2022 SKIP - CompatTool - Nothing to do the '/home/user/.var/app/com.valvesoftware.Steam/.local/share/Steam/compatibilitytools.d/SteamTinkerLaunch/steamtinkerlaunch' symlink still points to '/app/utils/steamtinkerlaunch/bin/steamtinkerlaunch'
pid 5276 != 521, skipping destruction (fork without exec?)
pid 5273 != 521, skipping destruction (fork without exec?)
Gtk-Message: 02:16:45.032: Failed to load module "appmenu-gtk-module"
Gtk-Message: 02:16:45.075: Failed to load module "canberra-gtk-module"
Gtk-Message: 02:16:45.075: Failed to load module "canberra-gtk-module"
pid 5811 != 521, skipping destruction (fork without exec?)
pid 5809 != 521, skipping destruction (fork without exec?)
/app/utils/steamtinkerlaunch/bin/steamtinkerlaunch: line 13015: /home/user/.local/share/applications/ModOrganizer-steamtinkerlaunch-dl.desktop: No such file or directory
Wed Jul  6 02:16:45 AM EDT 2022 INFO - startMO2 - Starting '/home/user/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding/MO2/ModOrganizer.exe' in standalone mode with command:
Wed Jul  6 02:16:45 AM EDT 2022 INFO - startMO2 - WINEDEBUG="-all" WINEPREFIX="/home/user/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/mo2/compatdata/pfx" "" "/home/user/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding/MO2/ModOrganizer.exe"
```

**Comment by frostworx** [comment]
the little log snippet you pasted looks fine - so invalid.
I suggest reading the wiki

**Comment by cohesive-flight** [comment]
I've gone through both the Flatpak and MO2 pages of the wiki, but I don't see anything else I should be doing. All the other buttons on the main menu launch just fine, it's only MO2 that doesn't. The ModOrganizer.exe appears in the directory the log outputted, so I'm not sure what's preventing it from running.

**Comment by sonic2kk** [comment]
I don't think Flatpak supports MO2, it might be missing some dependencies for getting MO2 installed or allowing it to run. For example, Innoextract.

Also, in Flatpak, it can't create the `.desktop` files and have the Nexus integration. Look at this line:

```
/app/utils/steamtinkerlaunch/bin/steamtinkerlaunch: line 13015: /home/user/.local/share/applications/ModOrganizer-steamtinkerlaunch-dl.desktop: No such file or directory
```

I think as well that something is missing based on this line:

```
Wed Jul  6 02:16:45 AM EDT 2022 INFO - startMO2 - WINEDEBUG="-all" WINEPREFIX="/home/user/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/mo2/compatdata/pfx" "" "/home/user/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding/MO2/ModOrganizer.exe"
```

> but I don't see anything else I should be doing.

This might sadly be a limitation of using Flatpak, but I'll see if I can take a look at what that missing command is meant to be in the code. Maybe it's meant to be the path to the Proton version, which I'm pretty sure STL should be able to find.

**Comment by sonic2kk** [comment]
Looking at the code, it does indeed look like the `$MO2WINE` var isn't set. STL is trying to run MO2 but without setting the Wine/Proton version to run with. STL is meant to run MO2 with the version of Wine/Proton that you run the game with - Previously we used a different version but now at least by default we use the regular Proton game version, as relatively modern releases of MO2 and Proton play nice together.

Hopefully this won't be too hard to fix. Famous last words!

**Comment by sonic2kk** [comment]
Hmm, I'm not having a very easy time understanding the MO2 code, but I think the problem may lie somewhere in the `setMO2Vars` function which isn't correctly setting `MO2WINE` in Flatpak. And sadly I don't have Steam Flatpak on any systems where I could test.

Maybe it should correctly work in Flatpak but the function is falling over at this line:

```
/app/utils/steamtinkerlaunch/bin/steamtinkerlaunch: line 13015: /home/user/.local/share/applications/ModOrganizer-steamtinkerlaunch-dl.desktop: No such file or directory
```

Either way on Flatpak, we either need to put this in the correct folder, or we shouldn't try to set it at all if you can't set `.desktop` files from Flatpak (or we should give permission for the Flatpak to write to this folder?)

**Comment by frostworx** [comment]
I can't tell why I closed the issue before. I assume I haven't read that the attempt was within Flatpak and I probably concluded incorrectly that another person broke everything while messing around manually with the settings/compatdata and then blamed steamtinkerlaunch _(happened far too often already, I don't have patience and time for such requests)_
So please accept my apologies @cohesive-flight  and thanks for picking this up again @sonic2kk.
If it turns out that this is really a flatpak issue I'd suggest discussing it [here](https://github.com/frostworx/steamtinkerlaunch/issues/27), because more people who actually use flatpak read there :)
But before I'll try to find out what else could be wrong when I find the time - `$MO2WINE` indeed looks **very** suspisious 😀

**Comment by frostworx** [comment]
MO2WINE is still autogenerated fine here (not using flatpak). please provide a full log for further debugging

**Comment by cohesive-flight** [comment]
Unfortunately, I saw your reply shortly after I've uninstalled and reinstalled both Steam and STL, and now I'm receiving a different output altogether. I'll attach it here, alongside a short clip of what occurs whenever I attempt to launch MO2 from the main menu.

[steam-log.txt](https://github.com/frostworx/steamtinkerlaunch/files/9504761/steam-log.txt)

https://user-images.githubusercontent.com/106868963/188846933-798d3945-151f-43e1-9ea8-eb6a46c55bf5.mp4

**Comment by frostworx** [comment]
thanks for troubleshooting.
the stderr log unfortunately doesn't provide any usable information, so it would be great if you'd also run
`steamtinkerlaunch mo2 start` _(to directly start mo2 without opening the settings menu before)_
and append the log which is created under `/dev/shm/steamtinkerlaunch/steamtinkerlaunch.log`

**Comment by cohesive-flight** [comment]
~~I'm not sure how to pass an argument with a Flatpak, or directly run STL, since it doesn't appear when tabbing: `flatpak run com.valvesoftware.Steam.`.~~

~~I looked at the other Flatpak thread to see if I could learn how, and someone mentioned: `flatpak run com.valvesoftware.Steam --command ="steamtinkerlaunch" settings`, but that only launched Steam.~~

Figured it out! I can use the command:`flatpak run --command=bash com.valvesoftware.Steam` which will then allow me to use `steamtinkerlaunch mo2 start`.

It failed again, but hopefully this log can be more useful to you.

[steam.log](https://github.com/frostworx/steamtinkerlaunch/files/9509346/steam.log)
[steamtinkerlaunch.log](https://github.com/frostworx/steamtinkerlaunch/files/9509298/steamtinkerlaunch.log)

**Comment by frostworx** [comment]
Thanks!
Hm, is the path
`/home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/downloads/proton/custom/GE-Proton7-31.tar.gz`
supposed to be correct? _(so generally: does `/home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/` contain other steamtinkerlaunch files?)_

Also `/home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding/MO2/ModOrganizer.exe` indicates that `/home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/` could be wrong.

The attempt to autodetect some valid default proton version seems to fail as well, so I'd suggest you define a valid "Proton version"
in the "DEFAULT" menu and save it. This might fix/workaround the problem, but I'd guess that flatpak has a problem with finding _(or reading/writing in)_ `/home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/`.
Can't help with flatpak issues, so I'd suggest you ask [here](https://github.com/frostworx/steamtinkerlaunch/issues/27) if someone can help.

**Comment by cohesive-flight** [comment]
The ``/home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/`` path does contain other steamtinkerlauncher files such as collections, compatdata, downloads, etc.

I tried setting the Proton version in the "DEFAULT" menu to ``experimental-7.0-20220907`` and it didn't fix the issue, but the ``steamtinkerlaunch.log`` is now different, so I'll attach it here. There's also now a duplicate proton version listed in the "DEFAULT'" menu after trying to set it as the default and launching MO2.

I'll look into asking in the other Flatpak issue in a bit.

[steamtinkerlaunch-protondefault.log](https://github.com/frostworx/steamtinkerlaunch/files/9541150/steamtinkerlaunch-protondefault.log)

Edit: I accidentally misclicked and closed this issue. But this might be fine since I'll be moving into the other thread soon.

**Comment by sonic2kk** [comment]
@cohesive-flight I wonder if this was related to the MO2 downloading issue... Looking at this part of the log:

```
Sat Sep 10 01:55:02 PM EDT 2022 INFO - StatusWindow - for 'Installing 'ModOrganizer''
Sat Sep 10 01:55:02 PM EDT 2022 INFO - getLatestMO2Ver - Search for latest 'Mod.Organizer' Release
Sat Sep 10 01:55:02 PM EDT 2022 INFO - updateWinRes - Waiting for Window 'SteamTinkerLaunch-InstallMO2Status'
Sat Sep 10 01:55:02 PM EDT 2022 INFO - updateWinRes - Waiting for Window 'SteamTinkerLaunch-InstallMO2Status'
Sat Sep 10 01:55:03 PM EDT 2022 INFO - dlLatestMO2 - Downloading Mod.Organizer-2.4.4.exe to /home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/downloads/mo2 from 'https://github.com/ModOrganizer2/modorganizer/releases/download/v2.4.4/Mod.Organizer-2.4.4.exe'
Sat Sep 10 01:55:03 PM EDT 2022 INFO - notiShow - Message 'Downloading 'Mod.Organizer-2.4.4.exe'' should go to StatusWindow
Sat Sep 10 01:55:03 PM EDT 2022 INFO - dlCheck - Downloading 'Mod.Organizer-2.4.4.exe'
Sat Sep 10 01:55:03 PM EDT 2022 INFO - notiShow - Message 'Downloading '/home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/downloads/mo2/Mod.Organizer-2.4.4.exe'' should go to StatusWindow
Sat Sep 10 01:55:03 PM EDT 2022 INFO - dlCheck - 'wget -q --show-progress https://github.com/ModOrganizer2/modorganizer/releases/download/v2.4.4/Mod.Organizer-2.4.4.exe -O /home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/downloads/mo2/Mod.Organizer-2.4.4.exe'
Sat Sep 10 01:55:03 PM EDT 2022 INFO - notiShow - Message 'Download finished 'Mod.Organizer-2.4.4.exe'' should go to StatusWindow
Sat Sep 10 01:55:03 PM EDT 2022 ERROR - dlLatestMO2 - Download failed!
Sat Sep 10 01:55:03 PM EDT 2022 SKIP - installMO2 - 'Mod.Organizer-2.4.4.exe' not found - nothing to install - skipping
Sat Sep 10 01:55:03 PM EDT 2022 SKIP - listSteamLibraries - not recreating already available '/dev/shm/steamtinkerlaunch/SteamLibraries.txt'
Sat Sep 10 01:55:03 PM EDT 2022 SKIP - updateWinRes - Timeout waiting for Window 'SteamTinkerLaunch-InstallMO2Status'
Sat Sep 10 01:55:03 PM EDT 2022 INFO - updateWinRes - Waiting for Window 'SteamTinkerLaunch-InstallMO2Status'
Sat Sep 10 01:55:03 PM EDT 2022 INFO - setModGameSyms - Found game 'Skyrim Special Edition (489830)' with pfx '/path/to/library/steam/SteamLibrary/steamapps/compatdata/489830/pfx'
Sat Sep 10 01:55:03 PM EDT 2022 INFO - setModGameSyms - Game Dir is '/path/to/library/steam/SteamLibrary/steamapps/common/Skyrim Special Edition' - adding the windows variant into the registry, because it is required for this game
Sat Sep 10 01:55:03 PM EDT 2022 INFO - updateWinRes - Waiting for Window 'SteamTinkerLaunch-InstallMO2Status'
Sat Sep 10 01:55:04 PM EDT 2022 INFO - updateWinRes - Waiting for Window 'SteamTinkerLaunch-InstallMO2Status'
Sat Sep 10 01:55:04 PM EDT 2022 SKIP - updateWinRes - Timeout waiting for Window 'SteamTinkerLaunch-InstallMO2Status'
Sat Sep 10 01:55:04 PM EDT 2022 ERROR - prepareMO2 - No '/home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding/MO2/ModOrganizer.exe' found - can't continue
Sat Sep 10 01:55:04 PM EDT 2022 SKIP - startMO2 - Could not find '/home/user/.var/app/com.valvesoftware.Steam/config/steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding/MO2/ModOrganizer.exe' - can't start ModOrganizer - this should not happen
Sat Sep 10 01:55:05 PM EDT 2022 SKIP - updateWinRes - Timeout waiting for Window 'SteamTinkerLaunch-InstallMO2Status'
```

Looks similar to what I saw in #591.

This should be fixed in v11.11 and the Flatpak should be up-to-date. You should be able to update from the command like (not sure of the command - sorry!) or from your preferred Flatpak management frontend like GNOME Software or KDE Discover.

Please try again from there and see how you get on :smiley:

**Comment by cohesive-flight** [comment]
I tried it again with the latest version, and sadly it still doesn't launch, but the log does look a bit different now, so I'll attach it here.

[steamtinkerlaunchv2.log](https://github.com/frostworx/steamtinkerlaunch/files/9707519/steamtinkerlaunchv2.log)

**Comment by sonic2kk** [comment]
Hmm, the logs look very similar to me. The core issue is still that the mo2 downloaded check is failing, though I was able to run `wget -q --show-progress https://github.com/ModOrganizer2/modorganizer/releases/download/v2.4.4/Mod.Organizer-2.4.4.exe -O $HOME/Downloads/Mod.Organizer-2.4.4.exe'` just fine so I don't think this is a URL problem.

Could you try going to `$HOME/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch` and removing the `mo2` folder, then after this in the same `steamtinkerlaunch` directory could you go into the `downloads` folder and remove `mo2`? This will remove the previously downloaded MO2 and remove the previous MO2 compatdata. I'm thinking *maybe* this will prevent `wget` from failing.

If that path (`$HOME/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch`) doesn't exist, we may be onto something...

Out of curiosity, did you set the username in your logfile to `user`/is your username `user`? It's no problem if you did, basically I'm just wondering if maybe the path STL is trying to download the MO2 exe to is broken on Flatpak.

`wget` won't create the path with `-O` if it doesn't already exist, to my knowledge anyway. And if the paths are wrong here, this could be why it's failing. Though I'm not sure why it's wrong. I don't think MO2 is actually failing to download the executable, STL checks on a variable called `MO2SPATH` which is built from `"$MO2DLDIR/$MO2SETUP"`, where `$MO2DLDIR` is probably something like `$HOME/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/downloads/mo2` and `$MO2SETUP` is the installer executable that gets downloaded to that folder. STL outputs `Download failed` when this file isn't downloaded to that folder.

I highly doubt it but it's also possible Flatpak doesn't have wget, though I would be... surprised to say the least if this was the case. I think it would've caused problems before now and probably would've been caught during development.

So in short I think STL is downloading MO2 but either wget isn't putting it in the expected folder or the expected path doesn't actually exist and is different on Flatpak.

**Comment by cohesive-flight** [comment]
The path does exist! I removed both MO2 directories, but it still wouldn't launch. And yes, I did change the username in the logfile. 

It does seem like it's not downloading it properly since I can't locate the MO2 executable in either directory (`downloads/mo2` is empty and `steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding` doesn't exist).

**Comment by HanPrower** [comment]
```
flatpak run --command="wget" --file-forwarding com.valvesoftware.Steam -q --show-progress https://github.com/ModOrganizer2/modorganizer/releases/download/v2.4.4/Mod.Organizer-2.4.4.exe -O /home/hanfox/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/downloads/mo2/Mod.Organizer-2.4.4.exe
Unknown option 'show-progress'
```

Removing `--show-progress` allows the download to work.

The Flatpak includes wget2.0.1 (this is part of the environment, not something the Flatpak is installing itself):
```
GNU Wget2 2.0.1 - multithreaded metalink/file/website downloader

+digest +https +ssl/gnutls +ipv6 +iri +large-file +nls -ntlm -opie +psl -hsts
+iconv +idn2 +zlib -lzma +brotlidec +zstd -bzip2 -lzip +http2 +gpgme
```

I don't see `--show-progress` in the manpage: https://manpages.debian.org/testing/wget2/wget2.1.en.html

I do however see it as an option in wget1 installed on my host system (which is Arch BTW):
```
GNU Wget 1.21.3 built on linux-gnu.

-cares +digest -gpgme +https +ipv6 +iri +large-file -metalink +nls 
+ntlm +opie +psl +ssl/gnutls

...

wget -h | grep progress
       --progress=TYPE             select progress gauge type
       --show-progress             display the progress bar in any verbosity mode
```

**Comment by frostworx** [comment]
good catch, @HanPrower 👍
The above commit might fix the problem.
if possible please retest, @cohesive-flight

**Comment by sonic2kk** [comment]
Not sure if Flatpak has any kind of git release but it might be possible to manually edit that part of the STL script to test

**Comment by cohesive-flight** [comment]
Currently there's no update available. How'd I implement this fix manually?

**Comment by HanPrower** [comment]
There won't be an update to the Flatpak until it's confirmed working. No point in breaking Flatpak more with a bad commit.

The correct way to test Flatpaks is to build them using `flatpak-builder`, but as STL is simply a shell script you can modify it easily enough... I don't really recommend this, but at worse if you break it you'll need to reinstall the STL Flatpak, so it's not too scary...

1. Close Steam
2. Navigate to `/var/lib/flatpak/runtime/com.valvesoftware.Steam.Utility.steamtinkerlaunch/x86_64/stable/active/files/bin`
3. Backup/copy (don't move) the existing `steamtinkerlaunch` file
4. `sudo` open the `steamtinkerlaunch` file (not the backed up one) and copy/paste the contents from the file in this repository over it
5. Run Steam
6. Test MO2

**Comment by cohesive-flight** [comment]
The `/var/lib/flatpak/runtime/` directory doesn't seem to exist for me, but I found one that looks similar enough: `~/.local/share/flatpak/runtime/com.valvesoftware.Steam.Utility.steamtinkerlaunch/x86_64/stable/active/files/bin/`.

I updated the `steamtinkerlaunch` file with the latest commit, but it caused STL to not launch anymore. I was able to replace it with the original file while Steam was running and it started to launch again.

I've attached the log Steam outputted when trying to use the latest commit (STL's log was empty).

[steamv2.log](https://github.com/frostworx/steamtinkerlaunch/files/9733796/steamv2.log)

**Comment by HanPrower** [comment]
I set up a beta branch on the Flatpak which has the latest commit from this repo, which may help test it as I'm unsure if you got it to work correctly before.

To install it run the following after closing Steam:
`flatpak --user remote-add --if-not-exists flathub-beta https://flathub.org/beta-repo/flathub-beta.flatpakrepo`

`flatpak update --appstream`

`flatpak --user install flathub-beta com.valvesoftware.Steam.Utility.steamtinkerlaunch`

You'll know if it worked as in STL's main window the version number will change to one with a date in. Flatpak is usually pretty good at guessing to use a beta over stable if it's installed, from what I've seen without extra steps.

To uninstall basically just run:
`flatpak --user remote-delete flathub-beta`

**Comment by sonic2kk** [comment]
Copypaste from the generic Flatpak issue:

Just a heads up that some software center frontends like KDE Discover will list the Flathub Beta under the sources list, if Flathub Beta is enabled and if the app has a Beta available. Other software frontends might do this but I can vouch that KDE Discover does this (installed Firefox Flatpak Beta on my Steam Deck this way and other software this way on my desktop)

<hr>

I don't use the Flatpak build but thanks a ton for providing a beta build - Hopefully this info above helps users access it in a more friendly way. It also allows you to revert back more easily to the stable build, should anything go wrong (just switch the source and reinstall)

**Comment by cohesive-flight** [comment]
I've installed the beta version, and it does launch as normal now. 

When attempting to install MO2, the installation box doesn't appear multiple times anymore, and after a few minutes, there was a notification of the installation being complete. I checked if the MO2 executable was in the directory: `~/.var/app/com.valvesoftware.Steam/.config/steamtinkerlaunch/mo2/compatdata/pfx/drive_c/Modding/MO2`, and it was! So the download issue is fixed now, but when trying to run it, it fails to do so.

So we're back to square one when I first opened this issue, but this time I'll provide STL's log of when it fails instead of reinstalling and experiencing a download issue instead. 

[steamtinkerlaunchv4.log](https://github.com/frostworx/steamtinkerlaunch/files/9749735/steamtinkerlaunchv4.log)

Based on the log, it seems that the reason why it's failing to run is because it's not able to find to find Wine set for it: `ERROR - setModWine - MO2WINE was not found - can't continue`

@sonic2kk did mention this earlier on when examining MO2's code.

**Comment by frostworx** [comment]
So which proton version did you configure for MO2 (global settings) and is it still available on your system?

**Comment by cohesive-flight** [comment]
The proton version set for MO2 was "GE-Proton7-37". I don't recall doing this, but I might've tried testing if STL's download function was only failing with MO2 by trying other features that require a download, such as downloading a custom proton version. It seemed that it was successful, as there are files in the custom proton directory.

I was curious if downloading it in the beta version instead would fix this issue of it not being found by MO2, but unfortunately it didn't work. However, there was a checksum error when downloading it, so it's probably corrupted, which might be why MO2 isn't able to find it.

When choosing a proton version for MO2, there was another option available in the drop-down menu called "experimental-7.0-20221005b". Once I tried this, MO2 was able to successfully launch!

So the issue might be that MO2's proton option was to set to a version that was corrupted. 

I'll attach a log of when I try downloading it.

[steamtinkerlaunch-checksum_error.log](https://github.com/frostworx/steamtinkerlaunch/files/9767974/steamtinkerlaunch-checksum_error.log)

**Comment by frostworx** [comment]
yay
